### ISSUE TITLE
False positive in `Lint/UselessAssignment`

### ISSUE THREAD

**Issue by faraaz-deepsource** [rubocop/rubocop#11159]
I have a file where I assign values in arrays from a block (simplified for this issue):
```ruby
# frozen_string_literal: true

arr_one = []
arr_two = []
arr_three = []
arrs = []

def try_get(arg1, arg2, &block)
  puts arg1
  puts arg2

  block.call
end

arr_one = try_get(1, 2) do
  4 + 5 # has few function calls and eventually return a value
end

arr_two = try_get(3, 4) do
  2 + 3 # has few function calls and eventually return a value
end

arr_three = try_get(4, 5) do
  0 + 1 # has few function calls and eventually return a value
end

arrs = [*arrs, *arr_one, *arr_two, *arr_three]
pp arrs
```
and I'm running on it:

```bash
rubocop --parallel --debug --only Lint/UselessAssignment useless_var.rb
```
--------

## Expected behavior

It should not report offenses since I clearly am using these variables further.

## Actual behavior
It reports offenses for `Lint/UselessAssignment` even though I am further using the assigned variables; even the suggested variable doesn't make much sense.
```
For /Users/faraaz/poc: Default configuration from /Users/faraaz/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.38.0/config/default.yml
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /Users/faraaz/poc/useless_var.rb
Loading cache from /Users/faraaz/.cache/rubocop_cache/b94b65c991d4e8bd6e48890d77d76ebc33141082/74c9cb4d2be4f8858e67033cf55c223c35495363/a7926327ac60c3bc15845aa0e1e020cb08cac07b
W

Offenses:

useless_var.rb:3:1: W: Lint/UselessAssignment: Useless assignment to variable - arr_one. Did you mean arr_two?
arr_one = []
^^^^^^^
useless_var.rb:4:1: W: Lint/UselessAssignment: Useless assignment to variable - arr_two. Did you mean arr_one?
arr_two = []
^^^^^^^
useless_var.rb:5:1: W: Lint/UselessAssignment: Useless assignment to variable - arr_three. Did you mean arr_two?
arr_three = []
^^^^^^^^^

1 file inspected, 3 offenses detected
Finished in 0.11380500000086613 seconds
```

## Steps to reproduce the problem

As described above.

## RuboCop version

```
➜  poc rubocop -V                                                             
1.38.0 (using Parser 3.1.2.1, rubocop-ast 1.23.0, running on ruby 3.1.2) [x86_64-darwin21]
```

**Comment by tejasbubane** [comment]
@faraaz-deepsource Hi, the empty arrays are not used and those variables are getting reassigned with `try_get`, hence the offenses are expected. You can remove the first four lines and the behaviour remains unchanged.

If the variables are used before reassignment, the cop will not report offenses:

```ruby
# frozen_string_literal: true

arr_one = []
arr_two = []
arr_three = []
arrs = []

# This usage makes the above assignment meaningful
puts [arr_one, arr_two, arr_three, arrs].inspect

# Rest code below same as before
def try_get(arg1, arg2, &block)
....
end
```

However couple things which seem like an issue:
1. `arrs = []` is also useless but not reported.
2. `Did you mean` should not report other useless variables

**Comment by faraaz-deepsource** [comment]
ohhh okay I get it now. The arrs will get reassigned anyway so that's why its a useless assignment
